cranelift/filetests/filetests/egraph/remat.clif

```diff
@@ -0,0 +1,35 @@
+test optimize
+set opt_level=none
+set use_egraphs=true
+target x86_64
+
+function %f(i32) -> i32 {
+block0(v0: i32):
+    v1 = iconst.i32 42
+    v2 = iadd.i32 v0, v1
+    brnz v2, block1
+    jump block2
+
+block1:
+    v3 = iconst.i32 84
+    v4 = iadd.i32 v2, v3
+    return v4
+
+block2:
+    return v2
+}
+
+; check:  block0(v0: i32):
+; nextln:     v1 = iconst.i32 42
+; nextln:     v2 = iadd v0, v1
+; nextln:     brnz v2, block1
+; nextln:     jump block2
+; check:   block1:
+; nextln:     v5 = iconst.i32 126
+; nextln:     v6 = iadd.i32 v0, v5
+; nextln:     return v6
+; check:   block2:
+; nextln:     v3 = iconst.i32 42
+; nextln:     v4 = iadd.i32 v0, v3
+; nextln:     return v4
```

Comment:
This rematerialization test case seems like an odd example since `block2` is strictly worse on any machine that has... um... at least one register, right?

It's a nice example of re-associating and constant-folding, though.

Reply:
Right, it's testing specifically that remat is preferred even when GVN would do the opposite naturally. It looks silly here where register pressure is low, but it happened to help significantly with SpiderMonkey, so...

(The case where it's most interesting is when there are a whole bunch of values like `x+1`, `x+2`, ... rematerialized inside a loop or otherwise deeper into a function, and when these all would otherwise be hoisted by LICM or GVN; then we have just one register `x` live instead of all of the values)